### PR TITLE
[INLONG-2266][Feature]Support reporting metrics by audit-sdk in sort

### LINKED ISSUE #2266: [Feature]Support reporting metrics by audit-sdk in sort
### Description

_No response_

### Use case

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by chantccc**
### Title Name: [INLONG-XYZ][component] Title of the pull request

where *XYZ* should be replaced by the actual issue number.

Fixes #2266 

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Review by doleyzi — APPROVED** [review]

**Review by dockerzhang — APPROVED** [review]

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/deserialization/DeserializationSchema.java [hunk 1]
```diff
@@ -181,6 +228,8 @@ public void removeDataFlow(DataFlowInfo dataFlowInfo) throws Exception {
                 multiTenancyDeserializer.removeDataFlow(dataFlowInfo);
                 fieldMappingTransformer.removeDataFlow(dataFlowInfo);
                 recordTransformer.removeDataFlow(dataFlowInfo);
+
+                inLongGroupIdAndStreamIdMap.remove(dataFlowInfo.getId());
```
Comment (luchunliang) [inline inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/deserialization/DeserializationSchema.java:232]: Remove the map between dataFlowId and inlongGroupIdStreamId immediately when the configuration data remove DataFlowInfo, maybe miss some audit data.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/deserialization/DeserializationSchema.java [hunk 2]
```diff
@@ -122,7 +145,23 @@ public void processElement(
...
+                if (auditImp != null) {
+                    Pair<String, String> groupIdAndStreamId = inLongGroupIdAndStreamIdMap.getOrDefault(
+                            serializedRecord.getDataFlowId(),
+                            Pair.of("", ""));
+
+                    auditImp.add(
+                            Constants.METRIC_AUDIT_ID_FOR_INPUT,
+                            groupIdAndStreamId.getLeft(),
+                            groupIdAndStreamId.getRight(),
+                            sinkRecord.getTimestampMillis(),
```
Comment (luchunliang) [inline inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/deserialization/DeserializationSchema.java:159]: Please check that sinkRecord.getTimestampMillis() is the generated time of pulsar/tube Message or the logged time of user data or current time.
    public DeserializationResult<SerializedRecord> deserialize(@SuppressWarnings("rawtypes") Message message)
            throws IOException {
        final byte[] data = message.getData();
        return DeserializationResult.of(new SerializedRecord(dataFlowId, message.getEventTime(), data), data.length);
    }
            deserializer.flatMap(mixedRow, new CallbackCollector<>((row -> {
                // each tid might be associated with multiple data flows
                for (long dataFlowId : dataFlowIds) {
                    collector.collect(new Record(dataFlowId, System.currentTimeMillis(), row));
                }
            })));

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/hive/HiveMultiTenantWriter.java [hunk 3]
```diff
@@ -126,6 +147,20 @@ public void processElement(SerializedRecord serializedRecord, Context context,
...
+            if (auditImp != null) {
+                Pair<String, String> groupIdAndStreamId = inLongGroupIdAndStreamIdMap.getOrDefault(
+                        serializedRecord.getDataFlowId(),
+                        Pair.of("", ""));
+
+                auditImp.add(
+                        Constants.METRIC_AUDIT_ID_FOR_OUTPUT,
+                        groupIdAndStreamId.getLeft(),
+                        groupIdAndStreamId.getRight(),
+                        serializedRecord.getTimestampMillis(),
```
Comment (luchunliang) [inline inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/hive/HiveMultiTenantWriter.java:160]: ditto

**Review by luchunliang — APPROVED** [review]

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/deserialization/DeserializationSchema.java (hunk 2 above)
Comment (chantccc) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/deserialization/DeserializationSchema.java:159]: sinkRecord.getTimestampMillis() is event time extracted from pulsar/tube message

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/hive/HiveMultiTenantWriter.java (hunk 3 above)
Comment (chantccc) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/hive/HiveMultiTenantWriter.java:160]: sinkRecord.getTimestampMillis() is event time extracted from pulsar/tube message

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/deserialization/DeserializationSchema.java (hunk 1 above)
Comment (chantccc) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/deserialization/DeserializationSchema.java:232]: Once the dataflow is removed, data in memory will be discarded, audit metric is sent only after the correspond data is processed.

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/deserialization/DeserializationSchema.java (hunk 2 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/deserialization/DeserializationSchema.java:159]: It is different between the generated time of pulsar/tube Message,the logged time of user data, current time.
current time>the generated time of pulsar/tube Message>the logged time of user data in agent sdk.
The log time of audit interface is for calculating delay interval in every step.

**Review by healchow — APPROVED** [review]

File: inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/deserialization/DeserializationSchema.java (hunk 1 above)
Comment (luchunliang) [reply inlong-sort/sort-core/src/main/java/org/apache/inlong/sort/flink/deserialization/DeserializationSchema.java:232]: Then data size in hive is bigger than data size in audit metric, some data have been sent but they do not get the callback.
But the gap can be receivable.

**Pull request merged**